### PR TITLE
[-Wdeprecated-copy-with-dtor] Fix deprecated copy with dtor

### LINKED ISSUE #2062: [-Wdeprecated-copy-with-dtor] definition of implicit copy assignment operator is deprecated
With the latest clang update it seems that MIOpen is hit with the following warning as error in build process:

```
# clang --version
AMD clang version 16.0.0 (ssh://gerritgit/lightning/ec/llvm-project amd-mainline-open 23126 df03e317430e013f05093fef7112cb245b25dd1c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/rocm/llvm/bin
```

at least the following files need to be updated:
image

For example, the warning messages look like:
```
/MIOpen/src/include/miopen/performance_config.hpp:43:13: warning: definition of implicit copy assignment operator for 'PerfConfig' is deprecated because it has a user-declared destructor [-Wdeprecated-copy-with-dtor]
    virtual ~PerfConfig() = default;
            ^
/MIOpen/src/include/miopen/performance_config.hpp:53:8: note: in implicit copy assignment operator for 'miopen::solver::PerfConfig' first required here
struct PerfConfigBase : PerfConfig
       ^
/MIOpen/src/include/miopen/solver.hpp:270:8: note: in implicit copy assignment operator for 'miopen::solver::PerfConfigBase<miopen::solver::PerformanceConfigConvAsm3x3U>' first required here
struct PerformanceConfigConvAsm3x3U : PerfConfigBase<PerformanceConfigConvAsm3x3U>
       ^
/MIOpen/src/include/miopen/serializable.hpp:109:14: note: in implicit copy assignment operator for 'miopen::solver::PerformanceConfigConvAsm3x3U' first required here
        self = out;
             ^
/MIOpen/src/include/miopen/performance_config.hpp:62:26: note: in instantiation of function template specialization 'miopen::solver::SerDes<>::Deserialize<miopen::solver::PerformanceConfigConvAsm3x3U>' requested here
        return SerDes<>::Deserialize(static_cast<Derived&>(*this), s);
                         ^
/MIOpen/src/include/miopen/solver.hpp:277:38: note: in instantiation of member function 'miopen::solver::PerfConfigBase<miopen::solver::PerformanceConfigConvAsm3x3U>::Deserialize' requested here
    PerformanceConfigConvAsm3x3U() : PerformanceConfigConvAsm3x3U(-1, -1, -1) {}

```

### PR DESCRIPTION
**Pull request by junliume**
attempt to fix #2062

### CODE REVIEW

**Comment by junliume** [comment]
Request review from @atamazov @DrizztDoUrden @averinevg

File: src/include/miopen/execution_context.hpp [hunk 1]
```diff
@@ -85,9 +85,14 @@ struct ExecutionContext
     inline Handle& GetStream() const { return *stream; }
     inline void SetStream(Handle* stream_) { stream = stream_; }
 
-    ExecutionContext() = default;
     ExecutionContext(Handle* stream_) : stream(stream_) {}
-    virtual ~ExecutionContext() = default;
+
+    ExecutionContext() {}
```
Comment (DrizztDoUrden) [inline src/include/miopen/execution_context.hpp]: Does this produce a warning with `= default` in ctor?

Comment (junliume) [reply src/include/miopen/execution_context.hpp]: I have pushed this docker: `rocm/miopen:debug_2063` and so far no warning messages.

**Review by DrizztDoUrden — DISMISSED** [review]
lgtm

File: src/include/miopen/execution_context.hpp [hunk 2]
```diff
@@ -85,9 +85,14 @@ struct ExecutionContext
...
+    virtual ~ExecutionContext()               = default;
+    ExecutionContext(const ExecutionContext&) = default;
+    ExecutionContext(ExecutionContext&&)      = default;
+    ExecutionContext& operator=(const ExecutionContext&) = default;
+    ExecutionContext& operator=(ExecutionContext&&) = default;
```
Comment (averinevg) [inline src/include/miopen/execution_context.hpp]: This looks a bit overkill :smile: 
Is adding default copy assignment operator not enough to suppress the warnings?

File: src/include/miopen/performance_config.hpp [hunk 3]
```diff
@@ -40,7 +40,13 @@ namespace solver {
 
 struct PerfConfig
 {
-    virtual ~PerfConfig() = default;
+    PerfConfig() {}
```
Comment (averinevg) [inline src/include/miopen/performance_config.hpp]: Does the absence of this constructor also produce a warning?

File: src/include/miopen/solver.hpp [hunk 4]
```diff
@@ -75,7 +75,13 @@ const int wave_size = 64;
 /// There could be multiple solvers of the same algorithm for a problem config.
 struct SolverBase
 {
-    virtual ~SolverBase() = default;
+    SolverBase() {}
```
Comment (averinevg) [inline src/include/miopen/solver.hpp]: Ditto

File: src/include/miopen/performance_config.hpp (hunk 3 above)
Comment (JehandadKhan) [reply src/include/miopen/performance_config.hpp]: I am under the impression that simply removing the dtor spec should be sufficient to silence the warning

File: src/include/miopen/utility/transposing_solver.hpp [hunk 5]
```diff
@@ -77,7 +77,12 @@ struct TransposeInvokeParams : InvokeParams
...
-    virtual ~TransposePseudoSolver()                                        = default;
+    TransposePseudoSolver() {}
+    virtual ~TransposePseudoSolver()                    = default;
+    TransposePseudoSolver(const TransposePseudoSolver&) = default;
+    TransposePseudoSolver(TransposePseudoSolver&&)      = default;
+    TransposePseudoSolver& operator=(const TransposePseudoSolver&) = default;
+    TransposePseudoSolver& operator=(TransposePseudoSolver&&)               = default;
     virtual std::string GetTranspose() const                                = 0;
     virtual ConvSolution GetSolution(const ExecutionContext& ctx,
                                      const TransposeProblem& problem) const = 0;
```
Comment (averinevg) [inline src/include/miopen/utility/transposing_solver.hpp]:
```suggestion
    virtual ~TransposePseudoSolver()                    = default;
    ...
protected:
    TransposePseudoSolver() = default;
    TransposePseudoSolver(const TransposePseudoSolver&) = default;
```

**Review by averinevg — CHANGES_REQUESTED** [review]

File: src/include/miopen/performance_config.hpp [hunk 6]
```diff
@@ -40,7 +40,13 @@ namespace solver {
...
+
+    virtual ~PerfConfig()         = default;
+    PerfConfig(const PerfConfig&) = default;
+    PerfConfig(PerfConfig&&)      = default;
+    PerfConfig& operator=(const PerfConfig&) = default;
+    PerfConfig& operator=(PerfConfig&&) = default;
```
Comment (averinevg) [inline src/include/miopen/performance_config.hpp]:
```suggestion
    virtual ~PerfConfig()         = default;
    ...
protected:
    PerfConfig() = default;
    PerfConfig(const PerfConfig&) = default;
    PerfConfig& operator=(const PerfConfig&) = default;
```

File: src/include/miopen/execution_context.hpp (hunk 2 above)
Comment (averinevg) [inline src/include/miopen/execution_context.hpp]:
```suggestion

    ExecutionContext() = default;
    virtual ~ExecutionContext()               = default;
    ExecutionContext(const ExecutionContext&) = default;
```

**Review by averinevg — CHANGES_REQUESTED** [review]

File: src/include/miopen/utility/transposing_solver.hpp (hunk 5 above)
Comment (averinevg) [reply src/include/miopen/utility/transposing_solver.hpp]: It is enough to add protected default constructor and protected copy constructor

File: src/include/miopen/performance_config.hpp (hunk 6 above)
Comment (averinevg) [reply src/include/miopen/performance_config.hpp]: Here you can add protected default constructor, protected copy constructor and protected copy assignment operator

File: src/include/miopen/execution_context.hpp (hunk 2 above)
Comment (averinevg) [reply src/include/miopen/execution_context.hpp]: Here you can add public copy constructor only

**Comment by atamazov** [comment]
@averinevg 👍

File: src/include/miopen/solver.hpp [hunk 7]
```diff
@@ -75,7 +75,13 @@ const int wave_size = 64;
...
+
+    virtual ~SolverBase()         = default;
+    SolverBase(const SolverBase&) = default;
+    SolverBase(SolverBase&&)      = default;
+    SolverBase& operator=(const SolverBase&) = default;
+    SolverBase& operator=(SolverBase&&) = default;
```
Comment (averinevg) [inline src/include/miopen/solver.hpp]:
```suggestion
    virtual ~SolverBase() = default;
    ...
protected:
    SolverBase() = default;
    SolverBase(const SolverBase&) = default;
```
It is enough to add protected default constructor and protected copy constructor

**Review by averinevg — DISMISSED** [review]

**Comment by junliume** [comment]
@averinevg have you tested using the `rocm/miopen:debug_2063` docker? The current issue will not be shown in current CI docker.

**Comment by averinevg** [comment]
> @averinevg have you tested using the `rocm/miopen:debug_2063` docker? The current issue will not be shown in current CI docker.

@junliume I have tested on my PC with clang 15.0.7

**Comment by atamazov** [comment]
@averinevg You have to use the same docker image (rocm/miopen:debug_2063 in this case) to reproduce and fix tidy issues.

**Comment by junliume** [comment]
**CORRECTION**: opps this is introduced by cd449969621198ef4368b633d3be3c71457929d7
Please review the latest of this branch, build is passing on the indicated docker now

@averinevg Indeed please pull docker rocm/miopen:debug_2063 and try to build MIOpen inside. I added your changes but now it is failing with multiple errors, for example:
```
In file included from /root/MIOpen/src/include/miopen/pooling/solvers.hpp:33:
/root/MIOpen/src/include/miopen/utility/transposing_solver.hpp:166:38: error: only virtual member functions can be marked 'override'
    std::string GetTranspose() const override { return "*-*"; }
                                     ^~~~~~~~~
/root/MIOpen/src/include/miopen/utility/transposing
```

**Comment by junliume** [comment]
Ping @averinevg @DrizztDoUrden @atamazov @JehandadKhan  for review since without it the build will fail in mainline soon.

File: src/include/miopen/execution_context.hpp [hunk 8]
```diff
@@ -85,9 +85,11 @@ struct ExecutionContext
     inline Handle& GetStream() const { return *stream; }
     inline void SetStream(Handle* stream_) { stream = stream_; }
 
-    ExecutionContext() = default;
     ExecutionContext(Handle* stream_) : stream(stream_) {}
-    virtual ~ExecutionContext() = default;
+
+    ExecutionContext()                        = default;
```
Comment (averinevg) [inline src/include/miopen/execution_context.hpp:90]: It's better not to move code that hasn't changed

**Review by averinevg — APPROVED** [review]
LGTM except one minor note

**Comment by averinevg** [comment]
> @averinevg You have to use the same docker image (rocm/miopen:debug_2063 in this case) to reproduce and fix tidy issues.

> @averinevg Indeed please pull docker rocm/miopen:debug_2063 and try to build MIOpen inside

@junliume @atamazov These warnings are reproducible using clang 15.0.7 as well

**Pull request merged**